tests/unit/machine/all_machines_unittest.cc

```diff
@@ -0,0 +1,384 @@
+{
+	auto fs = io::FileSystemRegistry::instance();
+	std::string class_name = machine->get_name();
+	fname = "shogun-unittest-AllMachines-trained_model_seiralization_consistency-" + class_name +
+			   ".XXXXXX";
+	generate_temp_filename(const_cast<char*>(fname.c_str()));
+
+	SG_REF(machine);
+	EXPECT_FALSE(fs->file_exists(fname));
+	std::unique_ptr<io::WritableFile> file;
+	EXPECT_FALSE(fs->new_writable_file(fname, &file));
+	auto fos = some<io::CFileOutputStream>(file.get());
+	auto serializer = some<io::CBitserySerializer>();
+	serializer->attach(fos);
+	serializer->write(wrap<CSGObject>(machine));
+}
+
+CMachine* deserialize_machine(std::string fname)
+{
+	auto fs = io::FileSystemRegistry::instance();
+	std::unique_ptr<io::RandomAccessFile> raf;
+	EXPECT_FALSE(fs->new_random_access_file(fname, &raf));
+	auto fis = some<io::CFileInputStream>(raf.get());
+	auto deserializer = some<io::CBitseryDeserializer>();
+	deserializer->attach(fis);
+	auto deser_obj = deserializer->read_object();
+	bool delete_success = !fs->delete_file(fname);
+	EXPECT_TRUE(delete_success);
+
+	return dynamic_cast<CMachine*>(deser_obj.get());
+}
+
+// TODO, generate this automatically, like in trained_model_serialization
+std::set<string> all_machines = {"LibSVM", "Perceptron", "LibLinear",
+		"MulticlassLibLinear", "LinearRidgeRegression", "KNN",
+		"KernelRidgeRegression", "LibLinearRegression", "RandomForest"};
+
+TEST(AllMachines, train_uninitialized)
+{
+	std::set<std::string> ignores = {};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		EXPECT_THROW(machine->train(), ShogunException);
+	}
+}
+
+TEST(AllMachines, train_execute)
+{
+	std::set<std::string> ignores = {};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		init_machine(machine);
+		auto data = generate_data(machine);
+		machine->set_labels(data.second);
+		machine->train(data.first);
+	}
+}
+
+TEST(AllMachines, train_thread_consistency)
+{
+	std::set<std::string> ignores = {
+			"RandomForest" // segfault
+			};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		init_machine(machine);
+		auto machine2 = make_clone(machine);
+
+		auto data = generate_data(machine);
+
+		get_global_parallel()->set_num_threads(1);
+		machine->set_labels(data.second);
+		if (machine->has("seed"))
+			machine->put("seed", 1);
+		machine->train(data.first);
+		auto result_single = machine->apply(data.first);
+
+		init_machine(machine);
+		get_global_parallel()->set_num_threads(4);
+		machine2->set_labels(data.second);
+		if (machine2->has("seed"))
+			machine2->put("seed", 1);
+		machine2->train(data.first);
```

Comment:
same as below: set_labels, train, apply. could you use a function for this?
the seeding part could be fixed by having that function a 4th arg that has a default value of some special number that means no seeding...